Content.Server/Atmos/Piping/Binary/EntitySystems/GasDualPortVentPumpSystem.cs

```diff
@@ -0,0 +1,101 @@
+using System;
+using Content.Server.Atmos.EntitySystems;
+using Content.Server.Atmos.Piping.Binary.Components;
+using Content.Server.Atmos.Piping.Components;
+using Content.Server.Atmos.Piping.Unary.Components;
+using Content.Server.NodeContainer;
+using Content.Server.NodeContainer.Nodes;
+using Content.Shared.Atmos;
+using Content.Shared.Atmos.Visuals;
+using JetBrains.Annotations;
+using Robust.Server.GameObjects;
+using Robust.Shared.GameObjects;
+
+namespace Content.Server.Atmos.Piping.Binary.EntitySystems
+{
+    [UsedImplicitly]
+    public class GasDualPortVentPumpSystem : EntitySystem
+    {
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<GasDualPortVentPumpComponent, AtmosDeviceUpdateEvent>(OnGasDualPortVentPumpUpdated);
+            SubscribeLocalEvent<GasDualPortVentPumpComponent, AtmosDeviceDisabledEvent>(OnGasDualPortVentPumpDisabled);
+        }
+
+        private void OnGasDualPortVentPumpUpdated(EntityUid uid, GasDualPortVentPumpComponent vent, AtmosDeviceUpdateEvent args)
+        {
+            var appearance = vent.Owner.GetComponentOrNull<AppearanceComponent>();
+
+            if (vent.Welded)
+            {
+                appearance?.SetData(VentPumpVisuals.State, VentPumpState.Welded);
+                return;
+            }
+
+            appearance?.SetData(VentPumpVisuals.State, VentPumpState.Off);
+
+            if (!vent.Enabled)
+                return;
+
+            if (!ComponentManager.TryGetComponent(uid, out NodeContainerComponent? nodeContainer))
+                return;
+
+            if (!nodeContainer.TryGetNode(vent.InletName, out PipeNode? inlet)
+            || !nodeContainer.TryGetNode(vent.OutletName, out PipeNode? outlet))
+                return;
+
+            var environment = args.Atmosphere.GetTile(vent.Owner.Transform.Coordinates)!;
+
+            // We're in an air-blocked tile... Do nothing.
+            if (environment.Air == null)
+                return;
+
+            if (vent.PumpDirection == VentPumpDirection.Releasing)
+            {
+                appearance?.SetData(VentPumpVisuals.State, VentPumpState.Out);
+                var pressureDelta = 10000f;
+
+                if (vent.PressureChecks.HasFlag(DualPortVentPressureBound.ExternalBound))
+                    pressureDelta = MathF.Min(pressureDelta, (vent.ExternalPressureBound - environment.Air.Pressure));
+
+                if (vent.PressureChecks.HasFlag(DualPortVentPressureBound.InputMinimum))
+                    pressureDelta = MathF.Min(pressureDelta, (inlet.Air.Pressure - vent.InputPressureMin));
```

Comment:
I think it was PJB that told us don't use HasFlag for enums coz allocs.

![image](https://user-images.githubusercontent.com/31366439/125596651-1f722c66-d9ad-4f10-9a03-1df364367402.png)

Reply:
to be fair, they only allocate on DEBUG, but changed now